docs/upgrading_to_nameko_3.rst

```diff
@@ -0,0 +1,347 @@
+
+    class Service:
+
+        @consume(
+            queue=Queue(
+                exchange=config["MY_EXCHANGE"],
+                routing_key=config["MY_ROUTING_KEY"],
+                name=config["MY_QUEUE_NAME"]
+            ),
+            prefetch_count=config["MY_CONSUMER_PREFETCH_COUNT"]
+        )
+        def consume(self, payload):
+            pass
+
+As you can see having config available outside service container also helps with configuring entrypoints and dependency providers by passing options to their constructors:
+
+.. code-block:: python
+
+    from nameko import config
+    from nameko_sqlalchemy import Database
+
+    class Service:
+        db = Database(engine_options=config["DB_ENGINE_OPTIONS"])
+
+.. _cli-define-option:
+
+Declaration of config in the CLI
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Configuration can now be declared on the CLI, rather than only being loaded from a file. This is useful for setting configuration that is specific to the environment in which the service is running, such as the hostname of a database server.
+
+Define config entries on the fly when running a service or shell:
+
+.. code-block:: bash
+
+    $ nameko run service \
+    --define AMQP_URI=pyamqp://someuser:*****@somehost/ \
+    --define max_workers=1000 \
+    --define SLACK="{'TOKEN': '*****'}"
+
+
+Upgrade Guide
+-------------
+
+Upgrading services
+~~~~~~~~~~~~~~~~~~
+
+This guide covers the steps to modify an existing service so that it works with Nameko 3.
+
+Generally only three areas are affected:
+
+1. :ref:`using-config-in-service-code`
+2. :ref:`using-config-in-tests`
+3. :ref:`using-custom-runners`
+
+.. _using-config-in-service-code:
+
+Using config in service code
+................................
+
+**Optional** — Use the global `config`` object rather than `Config`` dependency provider.
+The built-in `Config`` dependency provider from `nameko.dependency_providers` has been deprecated. You should instead use the `nameko.config` global.
+
+**Optional** — Use the global `config` object to simplify service setup.
+The global is available at all stages of the service lifecycle, including import time, which can make it much simpler to configure services.
+
+**Required ⚠** — Don't pass `config`` to `ServiceRpcProxy` or `ClusterRpcProxy`
+If you are using these standalone RPC clients from the `nameko.standalone` package, note that they no longer accept `config` as a parameter.
+
+You must set up config in the global context before using them, for example with `nameko.config.patch`.
+
+.. _using-config-in-tests:
+
+Using config in tests
+.....................
+
+**Required ⚠** — Read global `config` rather than any of the `x_config`` fixtures.
+There is a breaking change to the way that the `rabbit_config`, `test_config` and `empty_config` fixtures work. They no longer return the config dictionary.
+
+If you are using one of these fixtures, see the recommended pattern in the detailed breakdown of these configuration changes.
```

Comment:
Could we link to the "detailed breakdown" here so users can easily see what they're expected to do?